// eslint-plugin-regexp/dist/rules/prefer-w.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const regexp_ast_analysis_1 = require("regexp-ast-analysis");
function isSmallLetterRange(node) {
    return (node.type === "CharacterClassRange" &&
        node.min.value === utils_1.CP_SMALL_A &&
        node.max.value === utils_1.CP_SMALL_Z);
}
function isCapitalLetterRange(node) {
    return (node.type === "CharacterClassRange" &&
        node.min.value === utils_1.CP_CAPITAL_A &&
        node.max.value === utils_1.CP_CAPITAL_Z);
}
function isDigitRangeOrSet(node) {
    return ((node.type === "CharacterClassRange" &&
        node.min.value === utils_1.CP_DIGIT_ZERO &&
        node.max.value === utils_1.CP_DIGIT_NINE) ||
        (node.type === "CharacterSet" && node.kind === "digit"));
}
function isUnderscoreCharacter(node) {
    return node.type === "Character" && node.value === utils_1.CP_LOW_LINE;
}
exports.default = utils_1.createRule("prefer-w", {
    meta: {
        docs: {
            description: "enforce using `\\w`",
            category: "Stylistic Issues",
            recommended: true,
        },
        fixable: "code",
        schema: [],
        messages: {
            unexpected: "Unexpected {{type}} '{{expr}}'. Use '{{instead}}' instead.",
        },
        type: "suggestion",
    },
    create(context) {
        function createVisitor({ node, flags, getRegexpLocation, fixReplaceNode, getRegexpRange, fixerApplyEscape, toCharSet, }) {
            return {
                onCharacterClassEnter(ccNode) {
                    const charSet = toCharSet(ccNode);
                    let predefined = undefined;
                    if (charSet.equals(regexp_ast_analysis_1.Chars.word(flags))) {
                        predefined = "\\w";
                    }
                    else if (charSet.equals(regexp_ast_analysis_1.Chars.word(flags).negate())) {
                        predefined = "\\W";
                    }
                    if (predefined) {
                        context.report({
                            node,
                            loc: getRegexpLocation(ccNode),
                            messageId: "unexpected",
                            data: {
                                type: "character class",
                                expr: ccNode.raw,
                                instead: predefined,
                            },
                            fix: fixReplaceNode(ccNode, predefined),
                        });
                        return;
                    }
                    const lowerAToZ = [];
                    const capitalAToZ = [];
                    const digit = [];
                    const underscore = [];
                    for (const element of ccNode.elements) {
                        if (isSmallLetterRange(element)) {
                            lowerAToZ.push(element);
                            if (flags.ignoreCase) {
                                capitalAToZ.push(element);
                            }
                        }
                        else if (isCapitalLetterRange(element)) {
                            capitalAToZ.push(element);
                            if (flags.ignoreCase) {
                                lowerAToZ.push(element);
                            }
                        }
                        else if (isDigitRangeOrSet(element)) {
                            digit.push(element);
                        }
                        else if (isUnderscoreCharacter(element)) {
                            underscore.push(element);
                        }
                    }
                    if (lowerAToZ.length &&
                        capitalAToZ.length &&
                        digit.length &&
                        underscore.length) {
                        const unexpectedElements = [
                            ...new Set([
                                ...lowerAToZ,
                                ...capitalAToZ,
                                ...digit,
                                ...underscore,
                            ]),
                        ].sort((a, b) => a.start - b.start);
                        context.report({
                            node,
                            loc: getRegexpLocation(ccNode),
                            messageId: "unexpected",
                            data: {
                                type: "character class ranges",
                                expr: `[${unexpectedElements
                                    .map((e) => e.raw)
                                    .join("")}]`,
                                instead: "\\w",
                            },
                            *fix(fixer) {
                                const range = getRegexpRange(ccNode);
                                if (range == null) {
                                    return;
                                }
                                yield fixer.replaceTextRange(getRegexpRange(unexpectedElements.shift()), fixerApplyEscape("\\w"));
                                for (const element of unexpectedElements) {
                                    yield fixer.removeRange(getRegexpRange(element));
                                }
                            },
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
