// eslint-plugin-regexp/dist/rules/no-assertion-capturing-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const regexp_ast_analysis_1 = require("regexp-ast-analysis");
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-assertion-capturing-group", {
    meta: {
        docs: {
            description: "disallow capturing group that captures assertions.",
            category: "Possible Errors",
            recommended: true,
        },
        schema: [],
        messages: {
            unexpected: "Unexpected capture assertions.",
        },
        type: "suggestion",
    },
    create(context) {
        function createVisitor({ node, getRegexpLocation, }) {
            return {
                onCapturingGroupEnter(cgNode) {
                    if (regexp_ast_analysis_1.isZeroLength(cgNode)) {
                        context.report({
                            node,
                            loc: getRegexpLocation(cgNode),
                            messageId: "unexpected",
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
