// eslint-plugin-regexp/dist/rules/no-empty-alternative.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-empty-alternative", {
    meta: {
        docs: {
            description: "disallow alternatives without elements",
            category: "Possible Errors",
            recommended: false,
            default: "warn",
        },
        schema: [],
        messages: {
            empty: "No empty alternatives. Use quantifiers instead.",
        },
        type: "problem",
    },
    create(context) {
        function createVisitor({ node, getRegexpLocation, }) {
            function verifyAlternatives(regexpNode) {
                if (regexpNode.alternatives.length >= 2) {
                    for (let i = 0; i < regexpNode.alternatives.length; i++) {
                        const alt = regexpNode.alternatives[i];
                        if (alt.elements.length === 0) {
                            context.report({
                                node,
                                loc: getRegexpLocation(alt),
                                messageId: "empty",
                            });
                            return;
                        }
                    }
                }
            }
            return {
                onGroupEnter: verifyAlternatives,
                onCapturingGroupEnter: verifyAlternatives,
                onPatternEnter: verifyAlternatives,
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
