// eslint-plugin-regexp/dist/rules/control-character-escape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const CONTROL_CHARS = new Map([
    [0, "\\0"],
    [utils_1.CP_TAB, "\\t"],
    [utils_1.CP_LF, "\\n"],
    [utils_1.CP_VT, "\\v"],
    [utils_1.CP_FF, "\\f"],
    [utils_1.CP_CR, "\\r"],
]);
exports.default = utils_1.createRule("control-character-escape", {
    meta: {
        docs: {
            description: "enforce consistent escaping of control characters",
            category: "Best Practices",
            recommended: false,
        },
        fixable: "code",
        schema: [],
        messages: {
            unexpected: "Unexpected control character escape '{{actual}}' ({{cp}}). Use '{{expected}}' instead.",
        },
        type: "suggestion",
    },
    create(context) {
        function createVisitor({ node, getRegexpLocation, fixReplaceNode, }) {
            return {
                onCharacterEnter(cNode) {
                    if (cNode.parent.type === "CharacterClassRange") {
                        return;
                    }
                    const expectedRaw = CONTROL_CHARS.get(cNode.value);
                    if (expectedRaw === undefined) {
                        return;
                    }
                    if (cNode.raw === expectedRaw) {
                        return;
                    }
                    if (!utils_1.isRegexpLiteral(node) &&
                        cNode.raw === String.fromCodePoint(cNode.value)) {
                        return;
                    }
                    context.report({
                        node,
                        loc: getRegexpLocation(cNode),
                        messageId: "unexpected",
                        data: {
                            actual: cNode.raw,
                            cp: `U+${cNode.value
                                .toString(16)
                                .padStart(4, "0")}`,
                            expected: expectedRaw,
                        },
                        fix: fixReplaceNode(cNode, expectedRaw),
                    });
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
