// eslint-plugin-regexp/dist/rules/no-non-standard-flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const STANDARD_FLAGS = "gimsuy";
exports.default = utils_1.createRule("no-non-standard-flag", {
    meta: {
        docs: {
            description: "disallow non-standard flags",
            category: "Best Practices",
            recommended: false,
        },
        schema: [],
        messages: {
            unexpected: "Unexpected non-standard flag '{{flag}}'.",
        },
        type: "suggestion",
    },
    create(context) {
        function createVisitor({ node, getFlagsLocation, flagsString, }) {
            if (flagsString) {
                const nonStandard = [...flagsString].filter((f) => !STANDARD_FLAGS.includes(f));
                if (nonStandard.length > 0) {
                    context.report({
                        node,
                        loc: getFlagsLocation(),
                        messageId: "unexpected",
                        data: { flag: nonStandard[0] },
                    });
                }
            }
            return {};
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
