// eslint-plugin-regexp/dist/rules/no-useless-flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const ast_utils_1 = require("../utils/ast-utils");
const type_tracker_1 = require("../utils/type-tracker");
class RegExpReference {
    constructor(defineNode) {
        this.readNodes = new Map();
        this.state = {
            usedIn: {},
            track: true,
        };
        this.defineNode = defineNode;
    }
    addReadNode(node) {
        this.readNodes.set(node, {});
    }
    setDefineId(codePathId, loopNode) {
        this.defineId = { codePathId, loopNode };
    }
    markAsUsedInSearch(node) {
        const exprState = this.readNodes.get(node);
        if (exprState) {
            exprState.marked = true;
            exprState.usedInSearch = true;
        }
        this.state.usedIn.search = true;
    }
    markAsUsedInSplit(node) {
        const exprState = this.readNodes.get(node);
        if (exprState) {
            exprState.marked = true;
            exprState.usedInSplit = true;
        }
        this.state.usedIn.split = true;
    }
    markAsUsedInExec(node, codePathId, loopNode) {
        const exprState = this.readNodes.get(node);
        if (exprState) {
            exprState.marked = true;
            exprState.usedInExec = { id: { codePathId, loopNode } };
        }
        this.state.usedIn.exec = true;
    }
    markAsUsedInTest(node, codePathId, loopNode) {
        const exprState = this.readNodes.get(node);
        if (exprState) {
            exprState.marked = true;
            exprState.usedInTest = { id: { codePathId, loopNode } };
        }
        this.state.usedIn.test = true;
    }
    isUsed(kinds) {
        for (const kind of kinds) {
            if (this.state.usedIn[kind]) {
                return true;
            }
        }
        return false;
    }
    isCannotTrack() {
        return !this.state.track;
    }
    markAsUsed(kind) {
        this.state.usedIn[kind] = true;
    }
    markAsCannotTrack() {
        this.state.track = false;
    }
}
function getFlagLocation(context, node, flag) {
    const sourceCode = context.getSourceCode();
    if (node.type === "Literal") {
        const flagIndex = node.range[1] -
            node.regex.flags.length +
            node.regex.flags.indexOf(flag);
        return {
            start: sourceCode.getLocFromIndex(flagIndex),
            end: sourceCode.getLocFromIndex(flagIndex + 1),
        };
    }
    return node.arguments[1].loc;
}
function fixRemoveFlag({ flagsString, fixReplaceFlags }, flag) {
    if (flagsString) {
        return fixReplaceFlags(flagsString.replace(flag, ""));
    }
    return null;
}
function createUselessIgnoreCaseFlagVisitor(context) {
    return utils_1.defineRegexpVisitor(context, {
        createVisitor(regExpContext) {
            const { flags, regexpNode, toCharSet, ownsFlags } = regExpContext;
            if (!flags.ignoreCase || !ownsFlags) {
                return {};
            }
            const flagsNoI = Object.assign(Object.assign({}, flags), { ignoreCase: false });
            let unnecessary = true;
            return {
                onAssertionEnter(aNode) {
                    if (unnecessary) {
                        if (aNode.kind === "word" && flags.unicode) {
                            unnecessary = false;
                        }
                    }
                },
                onCharacterEnter(cNode) {
                    if (unnecessary) {
                        if (toCharSet(cNode).size > 1) {
                            unnecessary = false;
                        }
                    }
                },
                onCharacterSetEnter(cNode) {
                    if (unnecessary) {
                        if (cNode.kind === "word" && flags.unicode) {
                            unnecessary = false;
                        }
                        if (cNode.kind === "property") {
                            const caseInsensitive = toCharSet(cNode);
                            const caseSensitive = toCharSet(cNode, flagsNoI);
                            if (!caseInsensitive.equals(caseSensitive)) {
                                unnecessary = false;
                            }
                        }
                    }
                },
                onPatternLeave() {
                    if (unnecessary) {
                        context.report({
                            node: regexpNode,
                            loc: getFlagLocation(context, regexpNode, "i"),
                            messageId: "uselessIgnoreCaseFlag",
                            fix: fixRemoveFlag(regExpContext, "i"),
                        });
                    }
                },
            };
        },
    });
}
function createUselessMultilineFlagVisitor(context) {
    return utils_1.defineRegexpVisitor(context, {
        createVisitor(regExpContext) {
            const { flags, regexpNode, ownsFlags } = regExpContext;
            if (!flags.multiline || !ownsFlags) {
                return {};
            }
            let unnecessary = true;
            return {
                onAssertionEnter(node) {
                    if (node.kind === "start" || node.kind === "end") {
                        unnecessary = false;
                    }
                },
                onPatternLeave() {
                    if (unnecessary) {
                        context.report({
                            node: regexpNode,
                            loc: getFlagLocation(context, regexpNode, "m"),
                            messageId: "uselessMultilineFlag",
                            fix: fixRemoveFlag(regExpContext, "m"),
                        });
                    }
                },
            };
        },
    });
}
function createUselessDotAllFlagVisitor(context) {
    return utils_1.defineRegexpVisitor(context, {
        createVisitor(regExpContext) {
            const { flags, regexpNode, ownsFlags } = regExpContext;
            if (!flags.dotAll || !ownsFlags) {
                return {};
            }
            let unnecessary = true;
            return {
                onCharacterSetEnter(node) {
                    if (node.kind === "any") {
                        unnecessary = false;
                    }
                },
                onPatternLeave() {
                    if (unnecessary) {
                        context.report({
                            node: regexpNode,
                            loc: getFlagLocation(context, regexpNode, "s"),
                            messageId: "uselessDotAllFlag",
                            fix: fixRemoveFlag(regExpContext, "s"),
                        });
                    }
                },
            };
        },
    });
}
function createUselessGlobalFlagVisitor(context) {
    function reportUselessGlobalFlag(regExpReference) {
        const node = regExpReference.defineNode;
        context.report({
            node,
            loc: getFlagLocation(context, node, "g"),
            messageId: "uselessGlobalFlag",
        });
    }
    function isNeedReport(regExpReference) {
        let countOfUsedInExecOrTest = 0;
        for (const readData of regExpReference.readNodes.values()) {
            if (!readData.marked) {
                return false;
            }
            const usedInExecOrTest = readData.usedInExec || readData.usedInTest;
            if (usedInExecOrTest) {
                if (!regExpReference.defineId) {
                    return false;
                }
                if (regExpReference.defineId.codePathId ===
                    usedInExecOrTest.id.codePathId &&
                    regExpReference.defineId.loopNode ===
                        usedInExecOrTest.id.loopNode) {
                    countOfUsedInExecOrTest++;
                    if (countOfUsedInExecOrTest > 1) {
                        return false;
                    }
                    continue;
                }
                else {
                    return false;
                }
            }
        }
        return true;
    }
    return createRegExpReferenceExtractVisitor(context, {
        flag: "global",
        exit(regExpReferenceList) {
            for (const regExpReference of regExpReferenceList) {
                if (isNeedReport(regExpReference)) {
                    reportUselessGlobalFlag(regExpReference);
                }
            }
        },
        isUsedShortCircuit(regExpReference) {
            return regExpReference.isUsed([
                "match",
                "matchAll",
                "replace",
                "replaceAll",
            ]);
        },
    });
}
function createUselessStickyFlagVisitor(context) {
    function reportUselessGlobalFlag(regExpReference) {
        const node = regExpReference.defineNode;
        context.report({
            node,
            loc: getFlagLocation(context, node, "y"),
            messageId: "uselessStickyFlag",
        });
    }
    function isNeedReport(regExpReference) {
        for (const readData of regExpReference.readNodes.values()) {
            if (!readData.marked) {
                return false;
            }
        }
        return true;
    }
    return createRegExpReferenceExtractVisitor(context, {
        flag: "sticky",
        exit(regExpReferenceList) {
            for (const regExpReference of regExpReferenceList) {
                if (isNeedReport(regExpReference)) {
                    reportUselessGlobalFlag(regExpReference);
                }
            }
        },
        isUsedShortCircuit(regExpReference) {
            return regExpReference.isUsed([
                "search",
                "exec",
                "test",
                "match",
                "matchAll",
                "replace",
                "replaceAll",
            ]);
        },
    });
}
function createRegExpReferenceExtractVisitor(context, { flag, exit, isUsedShortCircuit, }) {
    const typeTracer = type_tracker_1.createTypeTracker(context);
    let stack = null;
    const regExpReferenceMap = new Map();
    const regExpReferenceList = [];
    function verifyForSearchOrSplit(node, kind) {
        const regExpReference = regExpReferenceMap.get(node.arguments[0]);
        if (regExpReference == null || isUsedShortCircuit(regExpReference)) {
            return;
        }
        if (!typeTracer.isString(node.callee.object)) {
            regExpReference.markAsCannotTrack();
            return;
        }
        if (kind === "search") {
            regExpReference.markAsUsedInSearch(node.arguments[0]);
        }
        else {
            regExpReference.markAsUsedInSplit(node.arguments[0]);
        }
    }
    function verifyForExecOrTest(node, kind) {
        const regExpReference = regExpReferenceMap.get(node.callee.object);
        if (regExpReference == null || isUsedShortCircuit(regExpReference)) {
            return;
        }
        if (kind === "exec") {
            regExpReference.markAsUsedInExec(node.callee.object, stack.codePathId, stack.loopStack[0]);
        }
        else {
            regExpReference.markAsUsedInTest(node.callee.object, stack.codePathId, stack.loopStack[0]);
        }
    }
    return utils_1.compositingVisitors(utils_1.defineRegexpVisitor(context, {
        createVisitor({ flags, regexpNode }) {
            if (flags[flag]) {
                const regExpReference = new RegExpReference(regexpNode);
                regExpReferenceList.push(regExpReference);
                regExpReferenceMap.set(regexpNode, regExpReference);
                for (const ref of ast_utils_1.extractExpressionReferences(regexpNode, context)) {
                    if (ref.type === "argument" || ref.type === "member") {
                        regExpReferenceMap.set(ref.node, regExpReference);
                        regExpReference.addReadNode(ref.node);
                    }
                    else {
                        regExpReference.markAsCannotTrack();
                    }
                }
            }
            return {};
        },
    }), {
        "Program:exit"() {
            exit(regExpReferenceList.filter((regExpReference) => {
                if (!regExpReference.readNodes.size) {
                    return false;
                }
                if (regExpReference.isCannotTrack()) {
                    return false;
                }
                if (isUsedShortCircuit(regExpReference)) {
                    return false;
                }
                return true;
            }));
        },
        onCodePathStart(codePath) {
            stack = {
                codePathId: codePath.id,
                upper: stack,
                loopStack: [],
            };
        },
        onCodePathEnd() {
            var _a;
            stack = (_a = stack === null || stack === void 0 ? void 0 : stack.upper) !== null && _a !== void 0 ? _a : null;
        },
        ["WhileStatement, DoWhileStatement, ForStatement, ForInStatement, ForOfStatement, " +
            ":matches(WhileStatement, DoWhileStatement, ForStatement, ForInStatement, ForOfStatement) > :statement"](node) {
            stack === null || stack === void 0 ? void 0 : stack.loopStack.unshift(node);
        },
        ["WhileStatement, DoWhileStatement, ForStatement, ForInStatement, ForOfStatement, " +
            ":matches(WhileStatement, DoWhileStatement, ForStatement, ForInStatement, ForOfStatement) > :statement" +
            ":exit"]() {
            stack === null || stack === void 0 ? void 0 : stack.loopStack.shift();
        },
        "Literal, NewExpression, CallExpression:exit"(node) {
            if (!stack) {
                return;
            }
            const regExpReference = regExpReferenceMap.get(node);
            if (!regExpReference || regExpReference.defineNode !== node) {
                return;
            }
            regExpReference.setDefineId(stack.codePathId, stack.loopStack[0]);
        },
        "CallExpression:exit"(node) {
            if (!stack) {
                return;
            }
            if (!ast_utils_1.isKnownMethodCall(node, {
                search: 1,
                split: 1,
                test: 1,
                exec: 1,
                match: 1,
                matchAll: 1,
                replace: 2,
                replaceAll: 2,
            })) {
                return;
            }
            if (node.callee.property.name === "search" ||
                node.callee.property.name === "split") {
                verifyForSearchOrSplit(node, node.callee.property.name);
            }
            else if (node.callee.property.name === "test" ||
                node.callee.property.name === "exec") {
                verifyForExecOrTest(node, node.callee.property.name);
            }
            else if (node.callee.property.name === "match" ||
                node.callee.property.name === "matchAll" ||
                node.callee.property.name === "replace" ||
                node.callee.property.name === "replaceAll") {
                const regExpReference = regExpReferenceMap.get(node.arguments[0]);
                regExpReference === null || regExpReference === void 0 ? void 0 : regExpReference.markAsUsed(node.callee.property.name);
            }
        },
    });
}
exports.default = utils_1.createRule("no-useless-flag", {
    meta: {
        docs: {
            description: "disallow unnecessary regex flags",
            category: "Best Practices",
            recommended: false,
            default: "warn",
        },
        fixable: "code",
        schema: [
            {
                type: "object",
                properties: {
                    ignore: {
                        type: "array",
                        items: {
                            enum: ["i", "m", "s", "g", "y"],
                        },
                        uniqueItems: true,
                    },
                },
                additionalProperties: false,
            },
        ],
        messages: {
            uselessIgnoreCaseFlag: "The 'i' flag is unnecessary because the pattern only contains case-invariant characters.",
            uselessMultilineFlag: "The 'm' flag is unnecessary because the pattern does not contain start (^) or end ($) assertions.",
            uselessDotAllFlag: "The 's' flag is unnecessary because the pattern does not contain dots (.).",
            uselessGlobalFlag: "The 'g' flag is unnecessary because not using global testing.",
            uselessStickyFlag: "The 'y' flag is unnecessary because not using sticky search.",
        },
        type: "suggestion",
    },
    create(context) {
        var _a, _b;
        const ignore = new Set((_b = (_a = context.options[0]) === null || _a === void 0 ? void 0 : _a.ignore) !== null && _b !== void 0 ? _b : []);
        let visitor = {};
        if (!ignore.has("i")) {
            visitor = utils_1.compositingVisitors(visitor, createUselessIgnoreCaseFlagVisitor(context));
        }
        if (!ignore.has("m")) {
            visitor = utils_1.compositingVisitors(visitor, createUselessMultilineFlagVisitor(context));
        }
        if (!ignore.has("s")) {
            visitor = utils_1.compositingVisitors(visitor, createUselessDotAllFlagVisitor(context));
        }
        if (!ignore.has("g")) {
            visitor = utils_1.compositingVisitors(visitor, createUselessGlobalFlagVisitor(context));
        }
        if (!ignore.has("y")) {
            visitor = utils_1.compositingVisitors(visitor, createUselessStickyFlagVisitor(context));
        }
        return visitor;
    },
});
