// eslint-plugin-regexp/dist/rules/prefer-t.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("prefer-t", {
    meta: {
        docs: {
            description: "enforce using `\\t`",
            category: "Stylistic Issues",
            recommended: true,
        },
        fixable: "code",
        schema: [],
        messages: {
            unexpected: "Unexpected character '{{expr}}'. Use '\\t' instead.",
        },
        type: "suggestion",
    },
    create(context) {
        function createVisitor(regexpContext, arrows) {
            const { node, getRegexpLocation, fixReplaceNode } = regexpContext;
            return {
                onCharacterEnter(cNode) {
                    if (cNode.value === utils_1.CP_TAB &&
                        cNode.raw !== "\\t" &&
                        !arrows.includes(cNode.raw)) {
                        context.report({
                            node,
                            loc: getRegexpLocation(cNode),
                            messageId: "unexpected",
                            data: {
                                expr: cNode.raw,
                            },
                            fix: fixReplaceNode(cNode, "\\t"),
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createLiteralVisitor(regexpContext) {
                return createVisitor(regexpContext, []);
            },
            createSourceVisitor(regexpContext) {
                return createVisitor(regexpContext, ["\t"]);
            },
        });
    },
});
